pygac_fdr/tests/test_end2end.py

```diff
@@ -0,0 +1,470 @@
+        filenames = []
+        for filename_gz in filenames_gz:
+            basename_gz = os.path.basename(filename_gz)
+            basename = os.path.splitext(basename_gz)[0]
+            filename = os.path.join(output_dir, os.path.basename(basename))
+            LOG.info('Decompressing {}'.format(filename_gz))
+            with gzip.open(filename_gz, 'rb') as f_in:
+                with open(filename, 'wb') as f_out:
+                    shutil.copyfileobj(f_in, f_out)
+            filenames.append(filename)
+        return filenames
+
+    @classmethod
+    def _run(cls, gac_files_gz, dbfile=None):
+        """Read GAC files and write calibrated scenes to netCDF.
+
+        If database file is given, also collect & update metadata.
+        """
+        if cls.resume:
+            # Resume testing with existing output files
+            LOG.info('Resuming test with existing output files')
+            return sorted(
+                glob.glob(os.path.join(cls.output_dir, '*.nc'))
+            )
+
+        # Prepare output directory
+        if not os.path.isdir(cls.output_dir):
+            os.makedirs(cls.output_dir)
+        cls._cleanup_output_dir()
+
+        # Unzip GAC files
+        gac_files = cls._unzip_gac_files(gac_files_gz, cls.output_dir)
+
+        # Read GAC files and write netCDF files
+        run = ['pygac-fdr-run',
+               '--cfg', cls.cfg_file,
+               '--output-dir', cls.output_dir,
+               '--tle-dir', cls.tle_dir,
+               '--verbose', '--log-all'] + gac_files
+        call_subproc(run)
+        nc_files = sorted(glob.glob(cls.output_dir + '/*.nc'))
+
+        if dbfile:
+            # Collect & complement metadata
+            collect = ['pygac-fdr-mda-collect', '--dbfile', dbfile, '--if-exists', 'replace',
+                       '--verbose'] + nc_files
+            call_subproc(collect)
+
+            # Update metadata
+            update = ['pygac-fdr-mda-update', '--dbfile', dbfile]
+            call_subproc(update)
+
+        return nc_files
+
+    @classmethod
+    def _cleanup_output_dir(cls):
+        if cls.cleanup:
+            for item in os.listdir(cls.output_dir):
+                item = os.path.join(cls.output_dir, item)
+                if os.path.isfile(item):
+                    os.unlink(item)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._cleanup_output_dir()
+
+    def _tst_regression(self, nc_files, nc_files_ref):
+        """Test entire netCDF contents against reference file."""
+        dynamic_attrs = ['date_created', 'history', 'version_satpy', 'version_pygac',
+                         'version_pygac_fdr']
+        for nc_file, nc_file_ref in zip(nc_files, nc_files_ref):
+            LOG.info('Performing regression test with {}'.format(nc_file))
+            with self.subTest(nc_file=nc_file):
+                with xr.open_dataset(nc_file) as ds:
+                    with xr.open_dataset(nc_file_ref) as ds_ref:
```

Comment:
See comment above on multiple with statements. In this case it also helps to avoid too many indentation levels.